### ISSUE TITLE
Kotlin Coding Practices

### ISSUE THREAD

**Issue by AllanWang** [TeamSubstratum/SubstratumThemeTemplate#47]
I'm glad to see that you guys have moved onto Kotlin, but a lot of the code still adopts a Java like structure. There are a few things I'm going to bring up for you to consider:

All links are based on the latest commit as of this issue

* [ThemeConstants](https://github.com/substratum/template/blob/4acc7d95d682ca60d88a31e6c212e3e5d5929343/app/src/main/java/substratum/theme/template/ThemerConstants.kt)

Why wrap constants in a singleton object? You can just add the values directly on the top level and also add the `const` keyword so that they are truly constants added where it is used in the byte code.

* [Substratum Launcher](https://github.com/substratum/template/blob/4acc7d95d682ca60d88a31e6c212e3e5d5929343/app/src/main/java/substratum/theme/template/SubstratumLauncher.kt#L58)

```kotlin
if (piracyChecker != null) piracyChecker!!.start()
```

should be

```kotlin
privacyChecker?.start()
```

Embrace the null operation rather than enforcing non nulls due to using `var`

The privacy checker as a whole probably shouldn't be nullable. Just instantiate it lazily so its not done when the check is disabled.

* System Information

```kotlin
fun checkNetworkConnection(): Boolean? {
    var isConnected = false
    try {
        val process = Runtime.getRuntime().exec("ping -c 1 www.google.com")
        val returnVal = process.waitFor()
        isConnected = returnVal == 0
    } catch (e: Exception) {
        // Suppress error
    }
    return isConnected
}
```

You can make this return a nonnull Boolean. At the same time, this has an O(1) retrieval time and no parameters. Kotlin's guidelines would suggest you use the var get() syntax but that is up to you. You also probably shouldn't use something like this to begin with especially if it is in the ui thread, but I didn't look any further into it.

A whole bunch of the other functions here all require Context, so use extension functions. And furthermore, none of these should be wrapped in a singleton

**Comment by nicholaschum** [comment]
You're correct @AllanWang, thank you for bringing this to our attention.

You are free to do a pull request that we will merge. As we are indeed mainly Java developers, adopting Kotlin will still retain some of our Java habits.

**Comment by AllanWang** [comment]
@nicholaschum I was thinking about a PR, which brings me back to something I mentioned a while back. Why not rewrite this to be like a library so others can add a dependency and extend one file? It beats them having to copy all of this manually themselves. Plus on your kotlin branch there are some meta data pointing to strings that don't exist in the library, and I'd need to know what to do with them

**Comment by nicholaschum** [comment]
@AllanWang We're very open with the template, if your PR is clearly documented we are willing to make a full rebase through your idea. I like it!

**Comment by AllanWang** [comment]
@nicholaschum Okay I'll try to make time to push something later today. A few things to note:

Your main purpose for this library is to be able to send an intent for Substratum. Recently (or since last time I checked) you've added a bunch of other intents, presumably to help against privacy. I'm not sure if they're implemented yet so unless there is more info I won't be adding it. 

As for the class, it's really only meant to launch a function, so I'm going to create an extension function which you'll see later. I'd say that the majority of themers have their custom ui anyways, so having the extension would make it easier to integrate.

Whether you'd also consider adding a simple UI class as another activity is up to you.

**Comment by AllanWang** [comment]
Can someone review [this](https://github.com/substratum/template/blob/4acc7d95d682ca60d88a31e6c212e3e5d5929343/app/src/main/java/substratum/theme/template/SubstratumLauncher.kt#L213) method? When is `incomplete` supposed to be changed?

**Comment by AllanWang** [comment]
I have a partial template available [here](https://github.com/AllanWang/substratum-template).

The takeaway is that there is a [clean file](https://github.com/AllanWang/substratum-template/blob/master/library/src/main/kotlin/substratum/theme/template/SubstratumLauncher.kt) with a config class, as well as an extension function for any activity to execute. This results in a setup [like so](https://github.com/AllanWang/substratum-template/blob/master/sample/src/main/kotlin/substratum/theme/sample/MainActivity.kt), where users can toggle any of the configurations to their desire.

Users will be able to install this as a dependency, and need no further connections with the template apart from the `launchSubstratum` extension.

I recommend the team review their current template first because
1. There are a few things that are definitively wrong
2. There are a few things that look like they are WIP
3. There are a few undocumented things that I don't really understand and hence haven't implemented yet.

And lastly, even if this is not accepted as a rebase, I recommend that the team

1. Try to prefix their respective strings with `substratum_` or something clean to differentiate what's in the template from what's in a themer's app.  Something like `resourcePrefix "substratum_"` can enforce that.
2. Avoid using appcompat styles when you are using a normal activity, especially since this is for lollipop (marshmallow really) and up.
3. Hopefully remove the need for an Activity in the substratum logic altogether. Any context can launch an intent to the substratum theme, and there's really no need for the template to control when a themer's own activity should finish.

**Comment by AllanWang** [comment]
And lastly, I'm not sure of the validity of these piracy checkers. I didn't  take a close look at the substratum app, and given your current modifications things are likely to change, but having a check just for when a user launches this app directly would not help when a user launches the theme through substratum's own interface.

**Comment by nicholaschum** [comment]
Yes, the `incomplete` variable looks redundant. I will take a look at that when I have more time.

**Recommendation 1 Response:**

1. If they are definitively wrong, you could indeed do a PR and fix what was wrong. The template is just a dynamic template of which plugs in to substratum. A full understanding of why we could easily just rebase off to another template is because of the certain responses it gives back once the AP check is complete.

2. Release 10 is a kotlin WIP template, it's still running on AS Canary Preview dependencies. As I said before, a PR could simply fix this instead of saying stuff is wrong.

3. There's a specific, intentionally undocumented portion of the template due to antipiracy and the way substratum handles launches theme intents. I would suggest further understanding how the full code works prior to trying to decipher what is going on, as the Substratum Launcher is just an antipiracy check for the Javier Santos library and then it quits itself after passing a returnIntent.

**Recommendation 2 Response:**

1.  Unnecessary and completely redundant for Strings

2. Check the minimum API, it's for N/API24+. We can use Theme.Material.Light.Dialog.Alert and this is advised for themers when they are interested to invert it or just completely get rid of AppCompat.

3.  Read Recommendation 1 Response (3)


**Final response 20 mins ago:**

You should take a look at the source code for substratum to understand the bridge between the template and the app. I shall tell you it's not a simple walk in the park but you'll be going through hurdles as we intentionally made it more confusing to understand. It's a fun challenge. 

But I believe that this conversation has run it's course as a full understanding of each aspect of the template including understanding the hooks is crucial for any rebase.

**Comment by AllanWang** [comment]
What I meant for appcompat for example is your [dialog style](https://github.com/substratum/template/blob/kt-n/app/src/main/res/values/dialog_styles.xml) for example which is based on appcompat.

I can understand that your goal for the main substratum app is to make it less straightforward, but what about this template? 

The [incomplete](https://github.com/substratum/template/blob/kt-n/app/src/main/java/substratum/theme/template/SubstratumLauncher.kt#L222) toggle under detectThemeReady is never changed, so I'm not sure what else to expect other than to say that something is wrong.

[This line](https://github.com/substratum/template/blob/088ff7d23b044bfc63e8113e68740ee58390bcbd/app/src/main/java/substratum/theme/template/internal/SystemInformation.kt#L34) looks like it serves no purpose, but I'm not sure if that's an accident or if the activity check is intended.

Your [ThemerConstants](https://github.com/substratum/template/blob/kt-n/app/src/main/java/substratum/theme/template/ThemerConstants.kt#L23) is named as if everything is final and shouldn't be modified, but some of them are clearly configs. What about `MINIMUM_SUBSTRATUM_VERSION`? Or is that defined by your team and changed when necessary? This is purely for improving the design structure, but it's an important part nonetheless.

Of course I could send a small PR fixing things the way I think it should be done, but it would have made much more sense if there was a discussion on the best practices for this template, which is presumably to be used by others with lower experience in development.

Regardless of why this discussion has now been terminated, I guess I'll wait to see how your team progresses and wait until things are stable before investing my time and reworking it myself.

**Comment by nicholaschum** [comment]
Yes, that Dialog Style was the one I referred to as I appended ".Dialog.Alert" onto the end of my Material Style I brought up.

I did mention that I will take a look at the `incomplete` variable, since I was actually not the one who committed it. The `pm` check was also redundant. I agree.

ThemerConstants is pretty much final everywhere, because those are just, as the name of the file suggests, constants for themers. So every single option is modifiable by the themer. That includes `MINIMUM_SUBSTRATUM_VERSION` as they can ensure that their template only works on a newer substratum with newer antipiracy updates.

I agree with you to discuss how we should clean up the template. I can reopen this issue as a gesture of goodwill since I was expecting the discussion to continue even with this issue closed. The template is indeed meant to hook into the complexion of substratum in the most simple way possible. All themers need to touch is ThemerConstants.kt and modify there. 

You've raised a valid point in redundant code and we will need to clean it up, so how about listing them all or doing a small temporary PR to fix the broken code?
